### ISSUE TITLE
Configurable authorization/token endpoints

### ISSUE THREAD

**Issue by fenekku** [inveniosoftware/invenio-oauthclient#263]
## Is your feature request related to a problem? Please describe.
Our provider has different endpoints for their "test" and "prod" installation. So in our local environment we should use their test and in our stage/prod environment their prod. However, there isn't a set of config variables in invenio-oauthclient that can be overridden to change the endpoints only. Raw dicts are being passed around so it's very hard to contextually replace without having to pass large one-line dict in environment variables.  

## Describe  the solution you'd like
Something like (big lines here)

``invenio.cfg``

```python
OAUTHCLIENT_REMOTE_APPS = dict(
    <my-app>=<my-app-module>.REMOTE_APP_CONF,
)
```

*deployment context*
```ini
# These override the endpoints defined in <my-app-module>.REMOTE_APP_CONF
OAUTHCLIENT_<MY-APP-UPPER-CASE>_TOKEN_URL="<my-deployment-context-url>" 
OAUTHCLIENT_<MY-APP-UPPER-CASE>_AUTHORIZATION_URL="<my-deployment-context-url>"
# ... 
```

But solution 2 of alternatives below would probably be my preferred at the cost of refactoring module / breaking changes.
 
## Describe alternatives you've considered
1- Homegrown generation of `invenio.cfg` depending on the deployment context. Tedious and onerous.
2- Custom `OAuthApp` that knows configuration variables to fetch (will probably have to do it this way depending on interfaces), but re-architecting invenio-oauthclient to have that approach by default would be nicest.

**Comment by ntarocco** [comment]
Usually, you will have different base URLs for test and prod installation, e.g. `oauth-test.mydomain.com` and `oauth.mydomain.com`.

Do you need customizing a specific URL instead?
This can be already achieved (even easier when [this](https://github.com/inveniosoftware/invenio-oauthclient/pull/288) will be merged):
```
github=github.REMOTE_APP
github["params"]["access_token_url"] = ...
```

Keycloak helper has already a specific endpoint customization: `OAUTHCLIENT_KEYCLOAK_USER_INFO_URL`.

**Comment by fenekku** [comment]
Not sure I follow (but that's probably because this is from a year and 4 months ago :laughing: ):

> Usually, you will have different base URLs for test and prod installation,

Yes, we have different domains for the 2, so... ? We still need to change full urls (containing domain) per deployment context. If the code was refactored to accept relative urls on one side and a shared domain on the other, we would still desire the ability to pass the domain as a config variable.


For 
```python
github = github.REMOTE_APP
github["params"]["access_token_url"] = ... 
```

I am assuming this is in `invenio.cfg`? At NU, we don't use a different `invenio.cfg` per deployment context. We only use different environment variables (via .env files). `invenio.cfg` is shared across contexts since it's a python file actually and we don't want to maintain multiple copies that are similar-but-different across our different environments. This goes to point 1- above.

> Keycloak helper has already a specific endpoint customization:

We don't use Keycloak at the university.

In any case, I just implemented what I suggested for us at NU in our own custom authentication package a year or so ago and it still works for us:

```python
class NussoOidcOAuthApp(OAuthRemoteApp):
    """Northwestern University Single-Sign-On OpenID Connect OAuth app."""

    def __init__(self, oauth, name, **params):
        """Constructor."""
        params["authorize_url"] = (
            oauth.app.config.get("NUSSO_OIDC_APP_AUTHORIZE_URL") or
            params["authorize_url"]
        )
        params["access_token_url"] = (
            oauth.app.config.get("NUSSO_OIDC_APP_ACCESS_TOKEN_URL") or
            params["access_token_url"]
        )
        super().__init__(oauth, name, **params)


REMOTE_APP = dict(
    remote_app=NussoOidcOAuthApp,
    # ...
)

```

Our package also deals with a bunch of authentication/migration needs we have too.

In conclusion, if no-one else wants direct config variables to be able to change OIDC urls, you can close this / we are not holding our breath over it.

**Comment by ntarocco** [comment]
In your `invenio.cfg`, you can simply do the following (we are doing this in our instances DEV/PROD):
```
_OAUTH_REMOTE_APP_COMMON = dict(
    base_url=os.environ.get(
        "OAUTH_CERN_OPENID_BASE_URL",
        "https://keycloak-qa.cern.ch/auth/realms/cern",
    ),
    access_token_url=os.environ.get(
        "OAUTH_CERN_OPENID_ACCESS_TOKEN_URL",
        "https://keycloak-qa.cern.ch/auth/realms/cern/"
        "protocol/openid-connect/token",
    ),
    authorize_url=os.environ.get(
        "OAUTH_CERN_OPENID_AUTHORIZE_URL",
        "https://keycloak-qa.cern.ch/auth/realms/cern/"
        "protocol/openid-connect/auth",
    ),
)

OAUTH_REMOTE_APP = copy.deepcopy(cern_openid.REMOTE_APP)
OAUTH_REMOTE_APP["params"].update(_OAUTH_REMOTE_APP_COMMON)
```

**Comment by ntarocco** [comment]
Closing, given that the requested feature is already available.
